pufferlib/ocean/trash_pickup/trash_pickup.h

```diff
@@ -0,0 +1,596 @@
+#define TRASH 1
+#define TRASH_BIN 2
+#define AGENT 3
+
+#define ACTION_UP 0
+#define ACTION_DOWN 1
+#define ACTION_LEFT 2
+#define ACTION_RIGHT 3
+
+#define AGENT_IS_CARRYING 1
+#define AGENT_IS_NOT_CARRYING 0
+
+#define LOG_BUFFER_SIZE 1024
+
+// Helper macro for 1D indexing
+#define INDEX(env, x, y) ((y) * (env)->grid_size + (x))
+
+typedef struct Log {
+    float episode_return;
+    float episode_length;
+    float trash_collected;
+} Log;
+
+typedef struct LogBuffer {
+    Log* logs;
+    int length;
+    int idx;
+} LogBuffer;
+
+// LogBuffer functions
+LogBuffer* allocate_logbuffer(int size) {
+    LogBuffer* logs = (LogBuffer*)calloc(1, sizeof(LogBuffer));
+    logs->logs = (Log*)calloc(size, sizeof(Log));
+    logs->length = size;
+    logs->idx = 0;
+    return logs;
+}
+
+void free_logbuffer(LogBuffer* buffer) {
+    free(buffer->logs);
+    free(buffer);
+}
+
+Log aggregate_and_clear(LogBuffer* logs) {
+    Log log = {0};
+    if (logs->idx == 0) {
+        return log;
+    }
+    for (int i = 0; i < logs->idx; i++) {
+        log.episode_return += logs->logs[i].episode_return;
+        log.episode_length += logs->logs[i].episode_length;
+        log.trash_collected += logs->logs[i].trash_collected;
+    }
+    log.episode_return /= logs->idx;
+    log.episode_length /= logs->idx;
+    log.trash_collected /= logs->idx;
+    logs->idx = 0;
+    return log;
+}
+
+void add_log(LogBuffer* logs, Log* log) {
+    if (logs->idx == logs->length) {
+        return;
+    }
+    logs->logs[logs->idx] = *log;
+    logs->idx += 1;
+}
+
+typedef struct {
+    int type;      // Entity type: EMPTY, TRASH, TRASH_BIN, AGENT
+    int index;     // Index in the positions array (-1 if not applicable)
+} GridCell;
+
+typedef struct {
+    // Interface for PufferLib
+    float* observations;
+    int* actions;
+    float* rewards;
+    unsigned char* dones;
+    LogBuffer* log_buffer;
+
+    int grid_size;
+    int num_agents;
+    int num_trash;
+    int num_bins;
+    int max_steps;
+    int current_step;
+
+    float positive_reward;
+    float negative_reward;
+    float total_episode_reward;
+
+    GridCell* grid; // 1D array for grid
+    int (*agent_positions)[2]; // Agent positions
+    int* agent_carrying; // Carrying status of agents
+    int (*trash_positions)[2];  // Positions of trash
+    int* trash_presence;        // Presence indicator for each piece of trash
+    int (*bin_positions)[2];    // Positions of trash bins
+
+    int obs_size;
+} CTrashPickupEnv;
+
+void compute_observations(CTrashPickupEnv* env) {
+    float* obs = env->observations;
+    float norm_factor = 1.0f / env->grid_size;
+
+    int obs_per_agent = env->obs_size / env->num_agents;
+
+    for (int agent_idx = 0; agent_idx < env->num_agents; agent_idx++) {
+        int obs_index = agent_idx * obs_per_agent;
+
+        // Add the observing agent's own position and carrying status
+        obs[obs_index++] = (float) (env->agent_positions[agent_idx][0]) * norm_factor;
```

Comment:
Normalizing coords is fine (required, actually) iff you actually need global coords. Why this vs. just a 2d local crop of tiles?

Reply:
Made 2D local crop (to be pushed).